Fix Issue 18675 - Missing toHash in std/experimental/checkedint.d

vlasebian opened the pull request:
This is a clean version of #6458

edi33416 commented on std/experimental/checkedint.d:
```diff
@@ -472,6 +472,48 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
 
+    // toHash
+    /**
+
```
Please delete empty line

edi33416 commented on std/experimental/checkedint.d:
```diff
@@ -472,6 +472,48 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
+    implement `hookToHash`, but it has state, a hash will be generated for
+    the `Hook` using the built-in function and it will be xored with the
+    hash of the `payload`. Otherwise, the hash of the payload is returned.
+
```
Please delete empty line

edi33416 commented on std/experimental/checkedint.d:
```diff
@@ -472,6 +472,48 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
+    hash of the `payload`. Otherwise, the hash of the payload is returned.
+
+    */
+    size_t toHash() const nothrow @safe
```
Please remove the function attributes as they should be inferred from `Hook` and/or `payload`

jercaianu replied:
`toHash` is used by the associative array and the hashing function has to be `const nothrow @safe`. This will not work with `aa`, unless we force the `Hook` to respect the same attributes.
Instead of letting `toHash` infer attributes from the `Hook`, I think it's better for `toHash` to force the `Hook` to be `const nothrow @safe`.

edi33416 replied:
Cool. Thank you for the insight. I hadn't look at `AA`. Let's leave this as it is then.

I think we should still add a ddoc unittest that builds an AA of Checked

edi33416 commented on std/experimental/checkedint.d:
```diff
@@ -3063,3 +3105,90 @@ version(unittest) private struct CountOpBinary
+            static size_t hashMask = 0xFFFF_0000_FFFF_0000;
+        }
+
+        static size_t hookToHash(T)(T payload)
```
Mark `hookToHash` as `@nogc nothrow @trusted` and then you can check that `toHash` correctly infers the attributes from his `Hook`. In order to check for the attributes add a `lambda` inside the assert, like so
`assert((() @nogc nothrow @safe => x.toHash() == x.toHash())());`

This defines a `lambda function` using the short syntax. You can read more about lambdas [here](https://dlang.org/spec/expression.html#lambdas)

edi33416 commented on std/experimental/checkedint.d:
```diff
@@ -3063,3 +3105,90 @@ version(unittest) private struct CountOpBinary
+
+    assert(checked!Hook1(c).toHash() != checked!Hook2(c));
+    assert(checked!Hook1(c).toHash() != checked!Hook1(d));
+
```
Could you split this in two unittests? Then, you can use the `///` syntax to add this unittest to the [docs](https://dlang.org/spec/ddoc.html#using_ddoc_to_generate_examples).

wilzbach commented on changelog/std-experimental-checkedint-toHash.dd:
```diff
@@ -0,0 +1,21 @@
+{
+    static size_t hookToHash(T)(T payload)
+    {
+        auto h = (() @trusted => typeid(payload).getHash(&payload))();
```
Not sure whether we want to advertise this `@trusted` in the public changelog.

wilzbach commented on std/experimental/checkedint.d:
```diff
@@ -472,6 +472,46 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
+        static size_t hashFunc(T)(T objToHash)
+        {
+            return (() @trusted => typeid(objToHash).getHash(&objToHash))();
+        }
```
Why not `hashOf`?

wilzbach commented on std/experimental/checkedint.d:
```diff
@@ -472,6 +472,46 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
+            }
+            else
+            {
+                return hashFunc(payload) ^ hashFunc(hook);
```
Not covered

wilzbach commented on std/experimental/checkedint.d:
```diff
@@ -472,6 +472,46 @@ if (isIntegral!T || is(T == Checked!(U, H), U, H))
+        }
+        else static if (hasMember!(typeof(payload), "toHash"))
+        {
+            return payload.toHash();
```
Not covered

wilzbach commented:
Mind the CI failures on Windows:

```
std\experimental\checkedint.d(492): Error: `@safe` function `std.experimental.checkedint.Checked!(long, Hook1).Checked.toHash` cannot call `@system` function `std.experimental.checkedint.__unittest_L3108_C9.Hook1.hookToHash!(const(long)).hookToHash`
```

vlasebian commented:
I split the unittest and added an entry for `hookToHash` in the Hook Table. Also I changed the function a bit, because I couldn't think of a case when the payload has its own `toHash` function implemented (because checkedint always wraps an integer) so if the hook does not implement `hookToHash` the built-in function will be used.

JackStouffer commented:
@vlasebian 

> because I couldn't think of a case when the payload has its own toHash function implemented (because checkedint always wraps an integer) 

This is not the case. `Checked!(Checked!int, ProperCompare)` is totally valid.

vlasebian commented:
I made some changes. Hope it's ok now.

n8sh commented:
Rebased & replaced a use of `TypeInfo.getHash` with `hashOf`.

n8sh commented:
Noting for the record that I disagree with this in concept (as discussed in the other PR) but there's nothing wrong with the implementation.
